server/mvcc/backend/hooks_test.go

```diff
@@ -0,0 +1,62 @@
+// Copyright 2021 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package backend_test
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"go.etcd.io/etcd/server/v3/mvcc/backend"
+	betesting "go.etcd.io/etcd/server/v3/mvcc/backend/testing"
+)
+
+func TestBackendPreCommitHook(t *testing.T) {
+	bucket := []byte("bucket")
+	key := []byte("key")
+
+	b, _ := betesting.NewTmpBackendCfg(t,
+		backend.BackendConfig{
+			BatchInterval: time.Hour,
+			Hooks: backend.NewHooks(func(tx backend.BatchTx) {
+				k, v := tx.UnsafeRange(bucket, key, nil, 1)
+				t.Logf("OnPreCommit executed: %v %v", k, v)
+				assert.Len(t, k, 1)
+				assert.Len(t, v, 1)
+				tx.UnsafePut(bucket, key, append(v[0], byte('c')))
+			}),
+		})
+
+	tx := b.BatchTx()
+	func() {
+		tx.Lock()
+		defer tx.Unlock()
+		tx.UnsafeCreateBucket(bucket)
+		tx.UnsafePut(bucket, key, []byte(">"))
+	}()
+	tx.Commit()
+
+	// Empty commit.
+	tx.Commit()
+
+	func() {
+		rtx := b.BatchTx()
+		rtx.Lock()
+		defer rtx.Unlock()
+		_, v := rtx.UnsafeRange(bucket, key, nil, 1)
+		assert.Len(t, v, 1)
+		assert.Equal(t, []byte(">ccc"), v[0])
```

Comment:
Could you clarify how this happens (ideally in test comments)? I see 1 explicit put of ">" and 2 commits that should trigger hook (adding 2x "c"). I don't see source of third "c".

Reply:
Good catch.  It was driven by '0' time BatchLimit.